### PR TITLE
Use Range.Desc to store ConfState.Nodes

### PR DESCRIPTION
**Pull request by bdarnell**
This ensures that group membership is preserved across restarts.

### CODE REVIEW

File: storage/raft_test.go [hunk 1]
```diff
@@ -35,66 +102,105 @@ func internalChangeReplicasArgs(newNodeID proto.NodeID, newStoreID proto.StoreID
...
+	if err := mtc.stores[1].ExecuteCmd(proto.Increment, incArgs, incResp); err != nil {
 		t.Fatal(err)
 	}
 
-	// Verify that the same data is available on the replica.
 	if err := util.IsTrueWithin(func() bool {
-		getArgs, getResp := getArgs([]byte("a"), rangeID, store2.StoreID())
-		if err := store2.ExecuteCmd(proto.Get, getArgs, getResp); err != nil {
+		getArgs, getResp := getArgs([]byte("a"), mtc.rangeID, mtc.stores[1].StoreID())
+		if err := mtc.stores[1].ExecuteCmd(proto.Get, getArgs, getResp); err != nil {
```
Comment (spencerkimball) [inline storage/raft_test.go]: Unless stores[1] is the leader, this only works because we're not yet returning the truth from Range.IsLeader().

Comment (bdarnell) [inline storage/raft_test.go]: True. We need a non-transactional local read for tests like this. I'll add a TODO.

Elsewhere in the test we assume writes are forwarded from followers to leaders, so if we get rid of that (and I still think we should keep it), the test will need a bit of restructuring.

File: storage/range.go [hunk 2]
```diff
@@ -1753,7 +1765,25 @@ func (r *Range) AdminMerge(args *proto.AdminMergeRequest, reply *proto.AdminMerg
...
-func (r *Range) InternalChangeReplicas(req *proto.InternalChangeReplicasRequest,
-	resp *proto.InternalChangeReplicasResponse) {
-	// TODO(bdarnell): apply the change to r.Desc.
+func (r *Range) InternalChangeReplicas(batch engine.Engine,
+	req *proto.InternalChangeReplicasRequest,
+	reply *proto.InternalChangeReplicasResponse) {
+	r.Lock()
+	defer r.Unlock()
+	// Apply the committed membership change to r.Desc.
+	// TODO(bdarnell): does this need special atomicity consideration, or will it be
```
Comment (spencerkimball) [inline storage/range.go]: What we do need a TODO for is updating the meta record(s) for the range with the new replica info.

It occurs to me that we might need to rework what we're doing here and instead make changing replicas part of a distributed txn.

The other mechanism which would work is to do exactly this, but set a dirty bit in the range descriptor and have the leader of the range update the meta transaction records asynchronously. It's fine if they lag a bit as clients are caching their values anyway and don't depend on them being fully up to date for correctness. Clients with stale info should get turned away from any range that they contact which ends up no longer being a member of the group.

File: storage/store.go [hunk 3]
```diff
@@ -1016,9 +1018,12 @@ func (s *Store) processRaft() {
 				nodeID, storeID := decodeRaftNodeID(e.NodeID)
 				cmd.RaftID = groupID
 				ok := cmd.Cmd.SetValue(&proto.InternalChangeReplicasRequest{
-					NodeID:     nodeID,
-					StoreID:    storeID,
-					ChangeType: changeTypeRaftToInternal[e.ChangeType],
+					// TODO(bdarnell): do we need to smuggle the original timestamp through raft?
```
Comment (spencerkimball) [inline storage/store.go]: I believe this will be correct. Original timestamp doesn't do you any good. The most up to date value here is much better.

**Comment by spencerkimball** [comment]
LGTM

**Pull request merged**